### ISSUE TITLE
Problem of vote overriding and key switching

### ISSUE THREAD

**Issue by chnejohnson** [appliedzkp/maci#415]
## TL;DR

The problem is that messages are processed in reverse order of being published. This brings about that any message with nonce 1 can override the previous message regardless of key switching or not.

## Collusion resistance argument

According to the scenario in [A Technical Introduction to MACI 1.0](https://medium.com/privacy-scaling-explorations/a-technical-introduction-to-maci-1-0-db95c3a9439a), the article illustrates the collusion resistance argument in section **2. System Overview - b. Vote Overriding and Public Key Switching**.

The core concept is that:
> A voter can tell a briber that they are voting for option A, but in reality they voted for option B.

And the example is as follows:
1. Bob signs up with public key 1
2. Bob then creates a command that contains — a vote for option A and public key 2
3. Bob signs this command with public key 1, the key he used to sign up
4. Bob encrypts this command into a message and submits it to the MACI smart contracts
5. The coordinator decrypts this message, and checks to ensure that the command is signed by Bob’s previous key — public key 1. This message is valid.
6. The coordinator then records Bob’s vote for option A and updates his public key to public key 2

> At this point, Bob has successfully voted for option A, and in order to override this vote must send in a new vote with a signature from public key 2. At this point, a briber now tries to get Bob to vote for option B:

1. Bob creates a command that contains — a vote for option B and public key 1
2. Bob signs this command with public key 1, encrypts the message and submits it to the MACI smart contracts
3. Bob shows the briber the decrypted message as proof of his vote for option B
4. The coordinator decrypts Bob’s message and sees that the signature does not match up with public key 2 — Bob’s previous key added in his previous message. Therefore this message is invalid and this vote is not counted in the final tally.
5. The briber has no way of knowing whether the vote was valid or invalid, and so is not incentivized to offer bribes to other users.

## Implementation
Let's implement the above scenario with MACI. Assume that Bob's initial keypair are **pk** (public key) and **sk** (private key), and Bob's new key are denoted as **npk** and **nsk**.

Now Bob published two messages to vote for option 1 with vote weight equal to 3, and vote for option 2 with vote weight equal to 2 as follows:

```
1. pk+sk => 1:3:2 (vote-option=1, vote-weight=3, nonce=2)
2. pk+sk => 2:2:1 (vote-option=2, vote-weight=2, nonce=1)
```

The order above is based on the order of messages published. So now the coordinator tally votes, and she can get the following result that option 1 gets 3 votes and option 2 gets 2 votes:

```
0:0 1:3 2:2 3:0 4:0 ...
```

At this time, a briber comes and tries to get Bob to vote for option 4 with vote wieght equal to 5, denoted as `4:5`

So, Bob wanted to secretly change his key, submit messages based on his true preferences by the new key, and submit an invalid message for showing the briber as proof of his vote for `4:5`.

Now, for changing the key and submitting true messages, Bob published messages to vote for `1:3` and `2:2` as above, but this time he wanted to add a vote for `3:3`, so...

```
1. npk+nsk 3:3:3
2. npk+nsk 2:2:2
3. npk+sk 1:3:1 (Notice that he changed his public key by signing with the old private key.)
```

Now the voting results are as follows:

```
0:0 1:3 2:2 3:3 4:0 ...
```

The last step is to submit an invalid message to vote the way the briber wants him to. So Bob should publish a message to vote for `4:5` by the old key which is invalid.

```
1. pk+sk 4:5:1
```

At this point, Bob can tell the briber that this on-chain message is the vote for `4:5` and Bob can sign again by his old key to show that the message he signed is the same as the message on chain.

According to the collusion resistance argument, this scenario should work and the expected voting results would be as follows, which implied that the invalid message for `4:5` will not be counted by the coordinator.

```
0:0 1:3 2:2 3:3 4:0 ...
```

But...in reality, the voting results will be...
```
0:0 1:3 2:0 3:0 4:5 ...
```

why?

## Reproduce

Run the following `test.sh` in maci/cli/

```bash
#!/bin/bash

set -e


node build/index.js deployVkRegistry && \
node build/index.js setVerifyingKeys -s 10 -i 1 -m 2 -v 2 -b 1 \
    -p ./zkeys/ProcessMessages_10-2-1-2_test.0.zkey \
    -t ./zkeys/TallyVotes_10-1-2_test.0.zkey \
    -k 0x8CdaF0CD259887258Bc13a92C0a6dA92698644C0

node build/index.js create \
    -r 0x8CdaF0CD259887258Bc13a92C0a6dA92698644C0 && \

node ./build/index.js deployPoll -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a \
    -pk macipk.c974f4f168b79727ac98bfd53a65ea0b4e45dc2552fe73df9f8b51ebb0930330 \
    -t 100 -g 25 -mv 25 -i 1 -m 2 -b 1 -v 2 && \

# Alice:
# Public key:  macipk.3e7bb2d7f0a1b7e980f1b6f363d1e3b7a12b9ae354c2cd60a9cfa9fd12917391
# Private key: macisk.fd7aa614ec4a82716ffc219c24fd7e7b52a2b63b5afb17e81c22fe21515539c

node ./build/index.js signup \
    -p macipk.3e7bb2d7f0a1b7e980f1b6f363d1e3b7a12b9ae354c2cd60a9cfa9fd12917391 \
    -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a && \

node build/index.js publish \
    -p macipk.3e7bb2d7f0a1b7e980f1b6f363d1e3b7a12b9ae354c2cd60a9cfa9fd12917391 \
    -sk macisk.fd7aa614ec4a82716ffc219c24fd7e7b52a2b63b5afb17e81c22fe21515539c \
    -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a \
    -i 1 -v 1 -w 3 -n 2 -o 0
node build/index.js publish \
    -p macipk.3e7bb2d7f0a1b7e980f1b6f363d1e3b7a12b9ae354c2cd60a9cfa9fd12917391 \
    -sk macisk.fd7aa614ec4a82716ffc219c24fd7e7b52a2b63b5afb17e81c22fe21515539c \
    -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a \
    -i 1 -v 2 -w 2 -n 1 -o 0


# Alice's new key:
# Public key:  macipk.ec1ced99098b6013584784a961eca1ea7de2a707f287bcbde6ccc18cbfce36ae
# Private key: macisk.3ce4d908c6023645e49d05fd1d1e4504eaf66304d28f0c65353ae96888d6193

node build/index.js publish \
    -p macipk.ec1ced99098b6013584784a961eca1ea7de2a707f287bcbde6ccc18cbfce36ae \
    -sk macisk.3ce4d908c6023645e49d05fd1d1e4504eaf66304d28f0c65353ae96888d6193 \
    -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a \
    -i 1 -v 3 -w 3 -n 3 -o 0

node build/index.js publish \
    -p macipk.ec1ced99098b6013584784a961eca1ea7de2a707f287bcbde6ccc18cbfce36ae \
    -sk macisk.3ce4d908c6023645e49d05fd1d1e4504eaf66304d28f0c65353ae96888d6193 \
    -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a \
    -i 1 -v 2 -w 2 -n 2 -o 0

node build/index.js publish \
    -p macipk.ec1ced99098b6013584784a961eca1ea7de2a707f287bcbde6ccc18cbfce36ae \
    -sk macisk.fd7aa614ec4a82716ffc219c24fd7e7b52a2b63b5afb17e81c22fe21515539c \
    -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a \
    -i 1 -v 1 -w 3 -n 1 -o 0

# submit invalid vote
node build/index.js publish \
    -p macipk.3e7bb2d7f0a1b7e980f1b6f363d1e3b7a12b9ae354c2cd60a9cfa9fd12917391 \
    -sk macisk.fd7aa614ec4a82716ffc219c24fd7e7b52a2b63b5afb17e81c22fe21515539c \
    -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a \
    -i 1 -v 4 -w 5 -n 1 -o 0



node build/index.js timeTravel -s 100 && \
node build/index.js mergeMessages -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a -o 0 && \
node build/index.js mergeSignups -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a -o 0 && \
node build/index.js genProofs -x 0xf204a4Ef082f5c04bB89F7D5E6568B796096735a \
    -sk macisk.49953af3585856f539d194b46c82f4ed54ec508fb9b882940cbe68bbc57e59e \
    -o 0 \
    -r ~/rapidsnark/build/prover \
    -wp ./zkeys/ProcessMessages_10-2-1-2_test \
    -wt ./zkeys/TallyVotes_10-1-2_test \
    -zp ./zkeys/ProcessMessages_10-2-1-2_test.0.zkey \
    -zt ./zkeys/TallyVotes_10-1-2_test.0.zkey \
    -t tally.json \
    -f proofs/

cat ./tally.json
```

**Comment by corydickson** [comment]
This might still be intended behavior because we assume that the bribe would occur for the first message. 

So in this example `4:5` would be Bob's true vote and would show the briber the signed message for `2:2`. Though to work off of this example the expectation would be to send N+1 messages and change their key again to have their intended vote counted.

Although, this does bring up a good point about when such a bribe may occur; for example the briber could enforce a contract where the message must be signed very close to the deadline thus preventing a subsequent key change.

**Comment by chnejohnson** [comment]
I probably understood, but I'm just curious about that whether this intended behavior could simplify the mechanism into the one that just allows voters to override votes in the voting period. So on the client side, I can regard maci as a machanism with the feature of vote overriding, so the key switching might be less important to the users, is that correct?

**Comment by corydickson** [comment]
The key switching is necessary from an implementation perspective when counting votes, does this answer your question?

**Comment by chnejohnson** [comment]
Ok, it makes sense. But how about the key selling problem? In this case, the original private key can override previous votes. So a voter can just give away her private key and then get bribed. But if key switching exactly transfers the vote right to another private key, whether it can prevent this kind of key selling problem?
